### PR TITLE
fix paint order of ink feature

### LINKED ISSUE #58665: FlexibleSpaceBar causes any action's InkWell to be drawn behind it
Internal: b/158153996

## Steps to Reproduce

1. Run the code below:
```
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: CustomScrollView(
        slivers: <Widget>[
          SliverAppBar(
            pinned: true,
            backgroundColor: Color.fromARGB(255, 0, 154, 254),
            expandedHeight: 250.0,
            actions: <Widget>[
              IconButton(
                icon: const Icon(Icons.add_circle),
                tooltip: 'First button',
                onPressed: () {},
              ),
            ],
            flexibleSpace: FlexibleSpaceBar(
              centerTitle: true,
              title: const Text('Ripple bug repro'),
              background: Container(
                color: Color.fromARGB(255, 0, 154, 54),
              ),
            ),
          ),
          SliverFillRemaining(
            child: Text('Hello world'),
          ),
        ],
      ),
    );
  }
}

```
2. Long-press the action button in the AppBar. Notice that the InkWell is not visible.
3. Change the Container's background color opacity to much lower
4. Long-press the action button in the AppBar. Notice that the InkWell is visible.

**Expected results:** The InkWell should have been visible in both cases.

**Actual results:** The InkWell seems to be drawn behind the FlexibleSpaceBar, hiding it as a result.

<details>
  <summary>Logs</summary>

```
```

```
Analyzing sliver_ripple_test...                                 
No issues found! (ran in 1.5s)
```

```
[✓] Flutter (Channel stable, v1.17.2, on Linux, locale en_AU.UTF-8)
    • Flutter version 1.17.2 at /opt/flutter
    • Framework revision 5f21edf8b6 (7 days ago), 2020-05-28 12:44:12 -0700
    • Engine revision b851c71829
    • Dart version 2.8.3

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /home/kernald/.android/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /home/kernald/.android/sdk
    • Java binary at: /home/kernald/.local/share/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6514223/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[!] Android Studio (version 4.0)
    • Android Studio at /home/kernald/.local/share/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6514223
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Community Edition (version 2019.3)
    • IntelliJ at /home/kernald/.local/share/JetBrains/Toolbox/apps/IDEA-C/ch-0/193.6911.18
    • Flutter plugin version 44.0.3
    • Dart plugin version 193.6494.35

[✓] Connected device (1 available)
    • F5321 • QV705FFG0B • android-arm64 • Android 8.0.0 (API 26)

! Doctor found issues in 1 category.

```

</details>

### PR DESCRIPTION
**Pull request by chunhtai**
## Description

This PR add an additional material layer in appbar to prevent the FlexibleSpace widget from swallowing the ink splash

## Related Issues
Fixes https://github.com/flutter/flutter/issues/58665

## Tests

I added the following tests:
see files

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 28 untriaged digest(s) on patchset 1.
Please triage them at https://flutter-gold.skia.org/cl/github/59108

**Comment by chunhtai** [comment]
@HansMuller this is ready for review

**Comment by chunhtai** [comment]
@HansMuller a friendly bump

File: packages/flutter/test/material/app_bar_test.dart [hunk 1]
```diff
@@ -1074,6 +1075,47 @@ void main() {
     expect(find.byIcon(Icons.menu), findsNothing);
   });
 
+  testWidgets('AppBar ink slash draw on the correct canvas', (WidgetTester tester) async {
```
Comment (HansMuller) [inline packages/flutter/test/material/app_bar_test.dart:1078]: slash => splash

File: packages/flutter/test/material/app_bar_test.dart [hunk 2]
```diff
@@ -1074,6 +1075,47 @@ void main() {
...
+    final Key key = UniqueKey();
```
Comment (HansMuller) [inline packages/flutter/test/material/app_bar_test.dart:1079]: Please add a comment above this line:
```
// This is a regression test for https://github.com/flutter/flutter/issues/58665
```

File: packages/flutter/lib/src/material/app_bar.dart [hunk 3]
```diff
@@ -641,7 +641,12 @@ class _AppBarState extends State<AppBar> {
         fit: StackFit.passthrough,
         children: <Widget>[
           widget.flexibleSpace,
-          appBar,
+          // Creates a material widget to prevent the flexibleSpace from swallow
+          // the ink splash effect.
+          Material(
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app_bar.dart:646]: Shouldn't we be using the Ink widget here?

https://api.flutter.dev/flutter/material/MaterialType-class.html

Comment (chunhtai) [reply packages/flutter/lib/src/material/app_bar.dart:646]: Ah nice

**Comment by chunhtai** [comment]
This is pr is not solving the actual problem

**Comment by chunhtai** [comment]
edit closed by mistake

File: packages/flutter/lib/src/material/app_bar.dart (hunk 3 above)
Comment (chunhtai) [reply packages/flutter/lib/src/material/app_bar.dart:646]: I took a look at the ink widget, and couldn't figure out how to apply it here. If i understand correctly, the Ink widget just make the splash go through the decoration of the Ink. but there is a stack on top of the material widget

Comment (HansMuller) [reply packages/flutter/lib/src/material/app_bar.dart:646]: You're right. I thought Ink would draw the splash on top of the Material's descendants, but that's not right. Your original approach was the right one after all, sorry about the detour.

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**